### PR TITLE
RDM-8842 Increment Assigned User Count of Cases when Assigned to Users

### PR DESCRIPTION
**Pull request by mattnayler**
### JIRA link (if applicable) ###

[RDM-8842](https://tools.hmcts.net/jira/browse/RDM-8842) _"Increment Assigned User Count of Cases when Assigned to Users"_

### Change description ###

Increment an organisation's Assigned User Count for a Case when Case-Roles are assigned to users via Add Case Assigned Users and Roles operation.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

**Comment by MSancaktutar** [comment]
@mattnayler  - Was this this PR meant to be reviewed? If so, please just resolve the conflicts with the base branch, and I can start reviewing it. Thanks.

File: src/aat/resources/features/F-105/F-105.feature [hunk 1]
```diff
@@ -210,3 +199,79 @@ Feature: F-105: Add Case-Assigned Users and Roles
...
+     And the request [is made from an authorised application, by Dil, with the Case ID of C1, User ID of Olawale, proper Case Role CR-2 and the Organisation ID of Olawale],
+     And it is submitted to call the [Add Case-Assigned Users and Roles] operation of [CCD Data Store Api],
+     Then a positive response is received,
+     And the response has all the details as expected,
+     And a call [to verify Olawale's reception of the role CR-2 over the case C1] will get the expected response as in [S-105.8842.2_Verify_Case_Roles_for_Case_C1],
+     And a call [to verify the count of users assigned to a case] will get the expected response as in [F-105_Verify_Counter_Unchanged].
+
+   # RDM-8842 AC-3
+   @S-105.8842.3
+   Scenario: No organisation ID is provided by the user
```
Comment (MSancaktutar) [inline src/aat/resources/features/F-105/F-105.feature:245]: Should have a more descriptive scenario title, telling what is specific to this scenario.

File: src/aat/resources/features/F-105/F-105.feature [hunk 2]
```diff
@@ -210,3 +199,79 @@ Feature: F-105: Add Case-Assigned Users and Roles
...
+     And a user [Olawale - with an active solicitor profile and valid User ID],
+     And a case [C1, which Richard has just] created as in [F-105_Case_Data_Create_C1],
+     And a successful call [to check the number of users having access to C1 in its supplementary data] as in [F-105_Prerequisite_Counter_Check_Call],
+     When a request is prepared with appropriate values,
+     And the request [is made from an authorised application, by Dil, with the Case ID of C1, User ID of Olawale, proper Case Role CR-1 and no Organisation ID],
+     And it is submitted to call the [Add Case-Assigned Users and Roles] operation of [CCD Data Store Api],
+     Then a positive response is received,
+     And the response has all the details as expected,
+     And a call [to verify Olawale's reception of the role CR-1 over the case C1] will get the expected response as in [S-105.8842.3_Verify_Case_Roles_for_Case_C1],
+     And a call [to verify the count of users assigned to a case] will get the expected response as in [F-105_Verify_Counter_Unchanged].
```
Comment (MSancaktutar) [inline src/aat/resources/features/F-105/F-105.feature:258]: could we be more specific in wording here, telling that the count will be zero?

File: src/main/java/uk/gov/hmcts/ccd/domain/model/std/CaseAssignedUserRoleRequest.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.ToString;
+
+import static com.fasterxml.jackson.annotation.JsonInclude.Include;
+
+@ToString
+@NoArgsConstructor
+@Getter
+public class CaseAssignedUserRoleRequest extends CaseAssignedUserRole {
```
Comment (MSancaktutar) [inline src/main/java/uk/gov/hmcts/ccd/domain/model/std/CaseAssignedUserRoleRequest.java:14]: The 2 class names don't have a parallelism in the semantics, breaching the expected connotation of an IS-A relation. The sentence `CaseAssignedUserRoleRequest` IS A `CaseAssignedUserRole` is technically correct, but linguistically wrong. Could you please propose a better fitting name for one or both of these to have a proper IS-A reading?

File: src/main/java/uk/gov/hmcts/ccd/domain/service/caseaccess/CaseAccessOperation.java [hunk 4]
```diff
@@ -84,25 +92,118 @@ public void updateUserAccess(CaseDetails caseDetails, CaseUser caseUser) {
...
 
     @Transactional
-    public void addCaseUserRoles(List<CaseAssignedUserRole> caseUserRoles) {
+    public void addCaseUserRoles(List<CaseAssignedUserRoleRequest> caseUserRoles) {
+
+        Map<Long, List<CaseAssignedUserRoleRequest>> cauRolesByCaseId = getMapOfCaseAssignedUserRolesByCaseId(caseUserRoles);
+
+        Map<String, Map<String, Object>> newUserCounts = getNewUserCountByCaseAndOrganisation(cauRolesByCaseId);
+
+        cauRolesByCaseId.forEach((caseId, roles) ->
```
Comment (MSancaktutar) [inline src/main/java/uk/gov/hmcts/ccd/domain/service/caseaccess/CaseAccessOperation.java:101]: in the `value` of an entry, the list is not that of roles. perhaps a name like requestedAssignments could read

File: src/main/java/uk/gov/hmcts/ccd/domain/service/caseaccess/CaseAccessOperation.java [hunk 5]
```diff
@@ -84,25 +92,118 @@ public void updateUserAccess(CaseDetails caseDetails, CaseUser caseUser) {
...
+            roles.forEach(caseUserRole ->
```
Comment (MSancaktutar) [inline src/main/java/uk/gov/hmcts/ccd/domain/service/caseaccess/CaseAccessOperation.java:102]: how about `requestedAssignment` here?

File: src/main/java/uk/gov/hmcts/ccd/domain/service/caseaccess/CaseAccessOperation.java [hunk 6]
```diff
@@ -84,25 +92,118 @@ public void updateUserAccess(CaseDetails caseDetails, CaseUser caseUser) {
...
+                caseUserRepository.grantAccess(caseId, caseUserRole.getUserId(), caseUserRole.getCaseRole()))
+        );
+
+        if (newUserCounts != null) {
```
Comment (MSancaktutar) [inline src/main/java/uk/gov/hmcts/ccd/domain/service/caseaccess/CaseAccessOperation.java:106]: newUserCounts returned shouldn't be nullable. If there is nothing to return, the method calculating this should return an empty map.

File: src/main/java/uk/gov/hmcts/ccd/domain/model/std/CaseAssignedUserRoleRequest.java (hunk 3 above)
Comment (mattnayler) [reply src/main/java/uk/gov/hmcts/ccd/domain/model/std/CaseAssignedUserRoleRequest.java:14]: As discussed: will rename as `CaseAssignedUserRoleWithOrganisation`

File: src/aat/resources/features/F-105/F-105.feature (hunk 2 above)
Comment (mattnayler) [reply src/aat/resources/features/F-105/F-105.feature:258]: As discussed: will update the spec so it reads "[to verify the count of users assigned to a case **has NOT changed**]"

File: src/main/java/uk/gov/hmcts/ccd/domain/service/caseaccess/CaseAccessOperation.java [hunk 7]
```diff
@@ -84,25 +92,121 @@ public void updateUserAccess(CaseDetails caseDetails, CaseUser caseUser) {
...
+    private Map<String, Map<String, Object>> getNewUserCountByCaseAndOrganisation(Map<Long, List<CaseAssignedUserRoleWithOrganisation>> cauRolesByCaseId) {
+        Map<String, Map<String, Object>> result = new HashMap<>();
+
+        Map<Long, List<CaseAssignedUserRoleWithOrganisation>> caseUserRolesWhichHaveAnOrgId = cauRolesByCaseId.entrySet().stream()
+            .collect(Collectors.toMap(Map.Entry::getKey, e -> e.getValue().stream()
+                // filter out no organisation_id
+                .filter(caseUserRole -> StringUtils.isNoneBlank(caseUserRole.getOrganisationId()))
+                .collect(Collectors.toList())))
+            // filter cases that have no remaining roles
+            .entrySet().stream().filter(e -> !e.getValue().isEmpty())
```
Comment (smathangi) [inline src/main/java/uk/gov/hmcts/ccd/domain/service/caseaccess/CaseAccessOperation.java:157]: May be it's better to extract these into separate small methods for better readability and unit testing?

Comment (MSancaktutar) [reply src/main/java/uk/gov/hmcts/ccd/domain/service/caseaccess/CaseAccessOperation.java:157]: I do agree. I'll add this to the content of the ticket for minor tweaks of this major work.

**Pull request merged**